service/interpreter/signalReceiver.go

```diff
@@ -1,23 +1,42 @@
 package interpreter
 
 import (
+	"strings"
+
 	"github.com/indeedeng/iwf/gen/iwfidl"
 	"github.com/indeedeng/iwf/service"
-	"strings"
 )
 
 type SignalReceiver struct {
 	// key is channel name
-	receivedSignals map[string][]*iwfidl.EncodedObject
-	provider        WorkflowProvider
-	logger          UnifiedLogger
+	receivedSignals      map[string][]*iwfidl.EncodedObject
+	failWorkflowByClient bool
+	provider             WorkflowProvider
 }
 
 func NewSignalReceiver(ctx UnifiedContext, provider WorkflowProvider) *SignalReceiver {
 	sr := &SignalReceiver{
-		provider:        provider,
-		receivedSignals: map[string][]*iwfidl.EncodedObject{},
+		provider:             provider,
+		receivedSignals:      map[string][]*iwfidl.EncodedObject{},
+		failWorkflowByClient: false,
 	}
+
+	provider.GoNamed(ctx, "fail-workflow-handler", func(ctx UnifiedContext) {
+		for {
+			ch := provider.GetSignalChannel(ctx, service.FailWorkflowSignalChanncelName)
+
+			val := service.FailWorkflowSignalRequest{}
+			err := provider.Await(ctx, func() bool {
+				return ch.ReceiveAsync(&val)
+			})
+			if err != nil {
+				// break the loop to prevent goroutine leakage
+				break
+			}
+			sr.failWorkflowByClient = true
```

Comment:
I think we can just break the loop after this is received. Since we are not interested into anymore fail request